internal/ptx/snowflake.go

```diff
@@ -2,20 +2,130 @@ package ptx
+}
+
+type snowflakeRendezvousMethodDomainFronting struct{}
+
+func (d *snowflakeRendezvousMethodDomainFronting) Name() string {
+	return "domain_fronting"
+}
+
+func (d *snowflakeRendezvousMethodDomainFronting) AMPCacheURL() string {
+	return ""
+}
+
+func (d *snowflakeRendezvousMethodDomainFronting) BrokerURL() string {
+	return "https://snowflake-broker.torproject.net.global.prod.fastly.net/"
+}
+
+func (d *snowflakeRendezvousMethodDomainFronting) FrontDomain() string {
+	return "cdn.sstatic.net"
+}
+
+// NewSnowflakeRendezvousMethodAMP is a rendezvous method that
+// uses the AMP cache to perform the rendezvous.
+func NewSnowflakeRendezvousMethodAMP() SnowflakeRendezvousMethod {
+	return &snowflakeRendezvousMethodAMP{}
+}
+
+type snowflakeRendezvousMethodAMP struct{}
+
+func (d *snowflakeRendezvousMethodAMP) Name() string {
+	return "amp"
+}
+
+func (d *snowflakeRendezvousMethodAMP) AMPCacheURL() string {
+	return "https://cdn.ampproject.org/"
+}
+
+func (d *snowflakeRendezvousMethodAMP) BrokerURL() string {
+	return "https://snowflake-broker.torproject.net/"
+}
+
+func (d *snowflakeRendezvousMethodAMP) FrontDomain() string {
+	return "www.google.com"
+}
+
+// ErrSnowflakeNoSuchRendezvousMethod indicates the given rendezvous
+// method is not supported by this implementation.
+var ErrSnowflakeNoSuchRendezvousMethod = errors.New("ptx: unsupported rendezvous method")
+
+// NewSnowflakeRendezvousMethod creates a new rendezvous method by name. We currently
+// support the following rendezvous methods:
+//
+// 1. "domain_fronting" uses domain fronting with the sstatic.net CDN;
+//
+// 2. "" means default and it is currently equivalent to "domain_fronting" (but
+// we don't guarantee that this default may change over time);
+//
+// 3. "amp" uses the AMP cache.
+//
+// Returns either a valid rendezvous method or an error.
+func NewSnowflakeRendezvousMethod(method string) (SnowflakeRendezvousMethod, error) {
+	switch method {
+	case "domain_fronting", "":
+		return NewSnowflakeRendezvousMethodDomainFronting(), nil
+	case "amp":
+		return NewSnowflakeRendezvousMethodAMP(), nil
+	default:
+		return nil, ErrSnowflakeNoSuchRendezvousMethod
+	}
+}
+
+// SnowflakeDialer is a dialer for snowflake. You SHOULD either use a factory
+// for constructing this type or set the fields marked as MANDATORY.
 type SnowflakeDialer struct {
-	// newClientTransport is an optional hook for creating
+	// RendezvousMethod is the MANDATORY rendezvous method to use.
+	RendezvousMethod SnowflakeRendezvousMethod
+
+	// newClientTransport is an OPTIONAL hook for creating
 	// an alternative snowflakeTransport in testing.
 	newClientTransport func(config sflib.ClientConfig) (snowflakeTransport, error)
 }
 
+// NewSnowflakeDialer creates a SnowflakeDialer with default settings.
+func NewSnowflakeDialer() *SnowflakeDialer {
+	return &SnowflakeDialer{
+		RendezvousMethod:   NewSnowflakeRendezvousMethodDomainFronting(),
+		newClientTransport: nil,
+	}
+}
```

Comment:
Minor but important aspect. I've changed the pattern with which we construct a `SnowflakeDialer` to require using a constructor _or_ explicitly setting a rendezvous method. (I could have made the rendezvous method optional, but TBH I am starting to think that using constructors in our code base reduces the amount of functions we need to write to deal with default cases, and this seems good in terms of easier testing - opinions?)

Reply:
I think this is a reasonable approach and I've seen ti quite commonly used in golang projects.

Reply:
Thanks!